server/sandbox_run_linux.go

```diff
@@ -37,6 +39,234 @@ import (
 )
 
+// addToMappingsIfMissing ensures the specified id is mapped from the host.
+func addToMappingsIfMissing(ids []idtools.IDMap, id int64) []idtools.IDMap {
+	firstAvailable := int(0)
+	for _, r := range ids {
+		if int(id) >= r.HostID && int(id) < r.HostID+r.Size {
+			// Already present, nothing to do
+			return ids
+		}
+		if r.ContainerID+r.Size > firstAvailable {
+			firstAvailable = r.ContainerID + r.Size
+		}
+	}
+	newMapping := idtools.IDMap{
+		ContainerID: firstAvailable,
+		HostID:      int(id),
+		Size:        1,
+	}
+	return append(ids, newMapping)
+}
+
+func (s *Server) configureSandboxIDMappings(mode string, sc *pb.LinuxSandboxSecurityContext) (*storage.IDMappingOptions, error) {
+	if mode == "" {
+		// No mode specified but mappings set in the config file, let's use them.
+		if s.defaultIDMappings != nil {
+			uids := s.defaultIDMappings.UIDs()
+			gids := s.defaultIDMappings.GIDs()
+			return &storage.IDMappingOptions{UIDMap: uids, GIDMap: gids}, nil
+		}
+		return nil, nil
+	}
+
+	// Ignore the annotation if not explicitly set in the config file.
+	if !s.config.AllowUsernsAnnotation {
+		return nil, nil
+	}
+
+	// expect a configuration like: private:uidmapping=0:1000:2000,2000:1000:2000;gidmapping=0:1000:4000,4000:1000:2000
+	parts := strings.SplitN(mode, ":", 2)
+
+	values := map[string]string{}
+	if len(parts) > 1 {
+		for _, r := range strings.Split(parts[1], ";") {
+			kv := strings.SplitN(r, "=", 2)
+			if len(kv) != 2 {
+				return nil, errors.Errorf("invalid argument: %q", r)
+			}
+			values[kv[0]] = kv[1]
+		}
+	}
+
+	_, uidMappingsPresent := values["uidmapping"]
+	_, gidMappingsPresent := values["uidmapping"]
+	// allow these options only if running as root
+	if uidMappingsPresent || gidMappingsPresent {
+		if sc.GetRunAsUser().Value != 0 {
+			return nil, errors.New("cannot use uidmapping or gidmapping if not running as root")
+		}
+	}
+
+	switch parts[0] {
+	case "auto":
+		const t = "true"
+		ret := &storage.IDMappingOptions{
+			AutoUserNs: true,
+		}
+		keepID := values["keep-id"] == t
+		mapToRoot := values["map-to-root"] == t
+		if keepID && mapToRoot {
+			return nil, errors.Errorf("cannot use both keep-id and map-to-root: %q", mode)
+		}
+		if v, ok := values["size"]; ok {
+			s, err := strconv.Atoi(v)
+			if err != nil {
+				return nil, err
+			}
+			ret.AutoUserNsOpts.Size = uint32(s)
+		} else {
+			ret.AutoUserNsOpts.Size = 65536
+		}
+		if v, ok := values["uidmapping"]; ok {
+			uids, err := idtools.ParseIDMap(strings.Split(v, ","), "UID")
+			if err != nil {
+				return nil, err
+			}
+			ret.AutoUserNsOpts.AdditionalUIDMappings = append(ret.AutoUserNsOpts.AdditionalUIDMappings, uids...)
+		}
+		if v, ok := values["gidmapping"]; ok {
+			gids, err := idtools.ParseIDMap(strings.Split(v, ","), "GID")
+			if err != nil {
+				return nil, err
+			}
+			ret.AutoUserNsOpts.AdditionalGIDMappings = append(ret.AutoUserNsOpts.AdditionalGIDMappings, gids...)
+		}
+		if sc.GetRunAsUser() != nil {
```

Comment:
How are we handling the case where the user name to run as is set in the Dockerfile?

Reply:
`setupContainerUser` in container_create.go has the code to combine and come up with a final list of uid, gid and groups. I think we should refactor that code to return those values and reuse it here.

Reply:
I don't think that can affect how we create the user namespace.  The user namespace is created for the POD so all the uids/gids must be already defined.  If a container is created later with a different uid/gid that is not part of the user namespace, then that cannot be handled.

I think this still requires some discussions and if we ccan come up with something better.

Reply:
Yeah, just realized that after our conversations today :) :+1: